micrometer-core/src/main/java/io/micrometer/core/instrument/step/StepMeterRegistry.java

```diff
@@ -102,4 +102,16 @@ protected DistributionStatisticConfig defaultHistogramConfig() {
                 .merge(DistributionStatisticConfig.DEFAULT);
     }
 
+    @Override
+    public void close() {
+        // Move clock to start of next step and stop it.
+        if (clock instanceof StoppableClock) {
+            long now = clock.wallTime();
```

Comment:
You need to publish metrics from the previous batch before incrementing the clock here, otherwise they will be lost. I would do inside the if:

if (config.enabled()) {
            publishSafely();
        }
        stop();

still call close on the super regardless after this executes.

Reply:
I get what you are saying. Basically, assuming the step as 60 seconds. If I close the registry in 1st second, there is a probability that I am still publishing values for the last step. In this scenario, just skipping publish should be sufficient. Since @jonatan-ivanov  said they are implementing a better fix for this issue, the smart thing for me is to wait until I hear back from them.

Reply:
Another thing is, the micrometer currently doesn't have a way not to publish duplicate data for StepRegistries. E.g: StepCounter holds a value of 50 for a minute and if publish is called again within the minute same 50 will be reported which is again an issue. This scenario I mentioned above might happen when closing registries and this is exactly one of the issues we are facing. This is not just about short-lived batches, but long-running processes can over-report the value since they are reporting the value for the previous minute. This can be addressed by adding a boolean in Pushregistry and a variable to keep track of the last published in the StepMeterRegistry.

Same as above will wait for @jonatan-ivanov to comment about the fix their team is working on. I have some more things that I am happy to commit on this PR if required.